File: 2021/2021W09/PreppindataW09.py
import pandas as pd
import re
import decimal

context = decimal.getcontext()
context.rounding = decimal.ROUND_HALF_UP

cust_info = pd.read_excel(pd.ExcelFile("F:/Data/Customer Information.xlsx"),sheet_name=0)
cust_info = pd.DataFrame([map(str.strip, x) for x in cust_info['IDs'].str.split(' ').values.tolist()])
cust_info = cust_info.melt(value_name='ID', var_name='ToDrop').drop('ToDrop', axis=1).dropna()
cust_info['Phone'] = cust_info['ID'].apply(lambda x: re.search('(\d{6})(?=,)',x).group(1)) 
cust_info['Area Code Key'] = cust_info['ID'].apply(lambda x: re.search('(?<=,)(\d{2}[A-Z])',x).group(1)) 
cust_info['Product ID'] = cust_info['ID'].apply(lambda x: re.search('([A-Z]+)$',x).group(1)) 
cust_info['Quantity'] = cust_info['ID'].apply(lambda x: int(re.search('(\d+)(?=-)',x).group(1))) 

area_code = pd.read_excel(pd.ExcelFile("F:/Data/Area Code Lookup.xlsx"),sheet_name=0,dtype=str)
area_code = area_code[~area_code['Area'].isin(['Clevedon', 'Fakenham', 'Stornoway'])]
area_code['Area Code Key'] = area_code.apply(lambda x: x['Code'][-2:]+x['Area'][0:1], axis=1)
area_code['Key Count'] = area_code['Area'].groupby(area_code['Area Code Key']).transform('count')

df = pd.merge(cust_info[cust_info['Quantity']>0], area_code[area_code['Key Count']==1], on='Area Code Key')

prod = pd.read_excel(pd.ExcelFile("F:/Data/Product Lookup.xlsx"),sheet_name=0)
prod['Price'] = prod['Price'].apply(lambda x: float(x[1:]))

final = pd.merge(df, prod, on='Product ID')
final['Revenue'] = final['Price']*final['Quantity']
final = final.groupby(['Area', 'Product Name'], as_index=False).agg({'Revenue':'sum'})
final['Revenue'] =final['Revenue'].apply(lambda x: int(round(decimal.Decimal(x), 0)))
final['Area Total'] = final['Revenue'].groupby(final['Area']).transform('sum')
final['Rank'] = final['Revenue'].groupby(final['Area']).rank(ascending=False).astype(int)
final['% of Total – Product'] = final.apply(lambda x: float(round(decimal.Decimal(x['Revenue']/x['Area Total']*100), 2)), axis=1)
final = final[['Rank', 'Area', 'Product Name', 'Revenue', '% of Total – Product']].copy()

File: 2021/2021W09/preppindataW09.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(splitstackshape)

round_half_up <- function (x, digits=0){
  posneg = sign(x)
  z = trunc(abs(x) * 10 ^ (digits + 1)) / 10
  z = floor(z * posneg + 0.5) / 10 ^ digits
  return(z)
  }

final <- read_excel("F:/Data/Customer Information.xlsx") %>%
  cSplit(., 'IDs', ' ') %>%
  pivot_longer(everything(), names_to = c(".value",NA), names_pattern = '(ID)(.*)', values_drop_na=TRUE) %>%
  mutate('Phone' = str_extract(`ID`,'(\\d{6})(?=,)'),
         'Area Code Key' = str_extract(`ID`,'(?<=,)(\\d{2}[A-Z])'),
         'Product ID' = str_extract(`ID`,'([A-Z]+)$'),
         'Quantity' = as.integer(str_extract(`ID`,'(\\d+)(?=-)'))) %>%
  filter(`Quantity`>0) %>%
  merge(.,read_excel("F:/Data/Area Code Lookup.xlsx") %>%
          filter(!(`Area` %in% c('Clevedon', 'Fakenham', 'Stornoway'))) %>%
          mutate('Area Code Key' = paste0(str_sub(`Code`,-2),str_sub(`Area`,1,1))) %>%
          group_by(`Area Code Key`) %>%
          filter(n()==1), by='Area Code Key') %>%
  merge(.,read_excel("F:/Data/Product Lookup.xlsx") %>%
          mutate('Price' = as.numeric(str_sub(`Price`,2))), by='Product ID') %>%
  group_by(`Area`,`Product Name`) %>%
  summarise('Revenue' = round_half_up(sum(`Price`*`Quantity`)),
            .groups='drop') %>%
  group_by(`Area`) %>%
  mutate('Rank' = dense_rank(desc(`Revenue`)),
         '% of Total ¡V Product' = round_half_up(`Revenue`/sum(`Revenue`)*100,2)) %>%
  select(c('Rank', 'Area', 'Product Name', 'Revenue', '% of Total ¡V Product'))
 
View(final)
